Site Switcher: update the new site button to show it at the bottom of the sidebar

sejas opened the pull request:
#### Proposed Changes

* Add new `transparent` style to `@automattic/components` to use it on the SiteSwitcher and the Sidebar.
* Add the new button styles to storybook.
* Add new eslint rule to package/components to fix an error on stories.
* Refactor Site Switcher `manage sites` button to use the new button style.
* ~Show always `add new site` button for all users.~ Keep the current behavior and show the button for users with only one site.
* Show this button at the bottom of the calypso sidebar, admin menu.
* Add plus icon for collapse mode, check the gif below.

#### Testing Instructions

* On an account with only one site. Or applying the following diff to remove this requirement:



<details>

  <summary> Diff to show always the button for testing purposes.  </summary>

```diff
diff --git a/client/my-sites/sidebar/add-new-site.jsx b/client/my-sites/sidebar/add-new-site.jsx
index 6492ddd5be..12e0dd534e 100644
--- a/client/my-sites/sidebar/add-new-site.jsx
+++ b/client/my-sites/sidebar/add-new-site.jsx
@@ -1,19 +1,13 @@
 import { Button } from '@automattic/components';
-import { useSelector, useDispatch } from 'react-redux';
+import { useDispatch } from 'react-redux';
 import TranslatableString from 'calypso/components/translatable/proptype';
 import { onboardingUrl } from 'calypso/lib/paths';
 import { recordTracksEvent } from 'calypso/state/analytics/actions';
-import { getCurrentUser } from 'calypso/state/current-user/selectors';
 import { setLayoutFocus } from 'calypso/state/ui/layout-focus/actions';
 
 export const AddNewSite = ( { title } ) => {
 	const reduxDispatch = useDispatch();
 
-	const visibleSiteCount = useSelector( getCurrentUser ).visible_site_count;
-	if ( visibleSiteCount > 1 ) {
-		return null;
-	}
-
 	const onClick = () => {
 		reduxDispatch( recordTracksEvent( 'calypso_add_new_wordpress_click' ) );
 		reduxDispatch( setLayoutFocus( 'content' ) );
```

</details>

* Open `/sites`
* Click on any page
* Observe the "Add New Site" is in the left Sidebar, and it appears at the bottom.
* Check also for mobile devices
* Click on `< Switch Site`
* Observe the `Manage Sites` button has not been modified.

#### Screenshots

**Sidebar on Desktop**

<img width="270" alt="67322-desktop" src="https://user-images.githubusercontent.com/779993/189122106-95a067d9-fb54-4bb7-a6f6-72d8713ff63a.png">



**Sidebar on Mobile**

<img width="378" alt="67322-mobile" src="https://user-images.githubusercontent.com/779993/189122139-5b9a6452-ecf4-49ef-9d1e-0ba7ad97b7f0.png">

**Collapse Mode**

![67322- Collapse mode](https://user-images.githubusercontent.com/779993/189208340-51cdde88-9ed0-4fad-bcb9-b7cd07bb968a.gif)



**Storybook**
<img width="1276" alt="67322-storybook" src="https://user-images.githubusercontent.com/779993/189121966-3cf2bf76-36d3-453d-bdb3-37fd11a21e8a.png">



#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

- Closes https://github.com/Automattic/wp-calypso/issues/67040

sejas committed:
AdminMenu: show always add new button

sejas committed:
AdminMenu: show add new site at the bottom

sejas committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into update/site-switcher-admin-menu-show-new-site-button-at-the-bottom

sejas committed:
Package Components: add transparent style to button for Site Switcher

sejas committed:
Site Switcher: refactor add new site button to use transparent Button

sejas committed:
Site Switcher: refactor site switcher manage sites to use transparent button

sejas committed:
Package Components: restore package.json dependencies

sejas committed:
Package Components: add eslint no-extraneous-dependencies rule

The file `packages/components/src/button/index.stories.jsx` was throwing the following error:

> '@storybook/addon-actions' should be listed in the project's dependencies. Run 'npm i -S @storybook/addon-actions' to add iteslintimport/no-extraneous-dependencies

But it shouldn't be needed since that dependency is included in the `@automattic/calypso-storybook` workspace.

sejas commented on packages/components/.eslintrc.js:
```diff
@@ -16,5 +17,16 @@ module.exports = {
 				'no-restricted-imports': 'off',
 			},
 		},
+		{
+			files: [ '**/*.stories.*' ],
+			rules: {
+				'import/no-extraneous-dependencies': [
+					'error',
+					{
+						packageDir: [ path.join( __dirname, '..', 'calypso-storybook' ) ],
+					},
+				],
+			},
+		},
```
I decided to add this new eslint rule to stories files because I was getting this error when importing `@storybook/addon-actions` on `packages/components/src/button/index.stories.jsx`
```
'@storybook/addon-actions' should be listed in the project's dependencies. Run 'npm i -S @storybook/addon-actions' to add iteslintimport/no-extraneous-dependencies
```

vykes-mac replied:
couldn't we add it to the `devDependencies` in package.json file ? would that solve the import warning?

vykes-mac replied:
would it make sense to add `@storybook/addon-actions` as a dev dependency in `packages/components/package.json` file? I see it as dependency in multiple places eg. `packages/search/package.json`

danielbachhuber replied:
@noahtallen @Automattic/team-calypso Do you have an opinion on this? ^

noahtallen replied:
> would it make sense to add @storybook/addon-actions as a dev dependency in packages/components/package.json file? I see it as dependency in multiple places eg. packages/search/package.json

If this solves the issue, I think that would be a better approach. Typically, we try to explicitly list every dependency at the project level. Yarn will hoist everything to the top-level anyways, but it helps when building projects independently (potentially for npm publishing), or if there was ever a need to split a project out.

sejas replied:
done!, I've removed the eslint rule and I've added the library to package.json 👌 

https://github.com/Automattic/wp-calypso/pull/67322/commits/2da3cb9e557b8920bb0aeddaac57ddfcb28ddedf

sejas commented:
@SaxonF, could you confirm the Button design is what's expected?

vykes-mac commented:
@sejas This can be done in another PR but just to point it out, collapsing the sidebar shows the add site button cut off.
![image](https://user-images.githubusercontent.com/47489215/189199041-28413e36-f3ef-4675-a9e5-e90bbd93cc2f.png)
seems the logic that handles that was inside the `SidebarItem` item component.

sejas committed:
Site Switcher: add collapse mode

sejas committed:
Site Switcher: add new site button a11y

sejas committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into update/site-switcher-admin-menu-show-new-site-button-at-the-bottom

sejas commented:
> @sejas This can be done in another PR but just to point it out, collapsing the sidebar shows the add site button cut off.
@vykes-mac Good Catch!, I added the collapse mode 👌 . Can you check it again?

zaguiini commented on packages/components/src/button/index.stories.jsx:
```diff
@@ -1,5 +1,8 @@
 import Button from '.';
+// Provide theme-default CSS variables from the Sidebar styles
+// eslint-disable-next-line no-restricted-imports
+import 'calypso/my-sites/sidebar/style.scss';
```
I don't think we should use Calypso imports in the Components package. How about only adding the necessary variables in this story's SCSS file?

vykes-mac replied:
I was wondering about this but because its in a story file I didn't give it much attention. it's seems like a circular dependency.

sejas replied:
Done! I've created a new CSS file with those variables. 
https://github.com/Automattic/wp-calypso/pull/67322/commits/582b1183c0bdbf88d5093ed8f05769f2fadb7f72

zaguiini commented:
There's this weird bottom padding:

![image](https://user-images.githubusercontent.com/26530524/189212176-a6b596ca-2c5e-44b6-8bdf-efedf62832de.png)

Should the margin on the bottom be the same as the one that's applied on the sides? I think it would be more visually consistent, but on the flip side, it will be right at the edge of the screen, so I'm not sure about its UX.

---

The `Add New Site` span element is still present when the sidebar is collapsed:

![image](https://user-images.githubusercontent.com/26530524/189212722-53c43b24-5adf-40f7-b33e-a916dee98491.png)

The side effect is that there's a hit slop, but only below the button.

danielbachhuber commented on client/components/site-selector/style.scss:
```diff
@@ -194,22 +194,6 @@
-		background: var( --color-sidebar-menu-hover-background );
-	}
-}
-
```
Nice! 🧹

danielbachhuber reviewed (commented):
Great start! It's already looking super sharp 😁

sejas commented:
> The `Add New Site` span element is still present when the sidebar is collapsed:
> 
> The side effect is that there's a hit slop, but only below the button.

I've discovered that this is the approach we are using for all menu items, so the screen readers still working.
We can remove that padding-bottom 👌

sejas committed:
Package Components: update button story adding theme-default css variables

sejas committed:
Package Components: restore eslint and add @storybook/addon-actions to package.json

sejas committed:
Site Switcher: remove extra bottom space for add site button on sidebar when collapsed

sejas committed:
Site Switcher: restore show Add New Site Button for users with only one site

sejas committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into update/site-switcher-admin-menu-show-new-site-button-at-the-bottom

sejas committed:
Site Switcher: restore SidebarSeparator

sejas commented:
@Automattic/build could you take another look?

danielbachhuber reviewed (commented):
Can we use the same margin from the bottom as the Site Switcher?

<img width="338" alt="image" src="https://user-images.githubusercontent.com/36432/189384762-805a8552-235e-493d-851f-d54f76c2f074.png">

<img width="375" alt="image" src="https://user-images.githubusercontent.com/36432/189384827-61c1cd96-aba9-4b1f-9c69-d37526f5e25c.png">

Also, in Chrome, it disappears entirely off the screen:

<img width="378" alt="image" src="https://user-images.githubusercontent.com/36432/189384928-797ed4e5-0a67-434a-ba38-34adf195f14c.png">

It seems like the `min-height: 80px` might be problematic?

<img width="1299" alt="image" src="https://user-images.githubusercontent.com/36432/189385094-afc560b2-2e3c-447c-aac2-e030c4f10e8c.png">

In collapsed view, the icon is white when I'd expect it to match the color of the other icons:

<img width="216" alt="image" src="https://user-images.githubusercontent.com/36432/189385241-a1a9af7a-96d3-438a-affd-4a94caeb06d0.png">

sejas committed:
Site Switcher: change the sidebar actions spacings

sejas committed:
Site Switcher: change the color for add new site button on collapsed sidebar

sejas commented:
@danielbachhuber 

> Can we use the same margin from the bottom as the Site Switcher?

Sure!, done.

> Also, in Chrome, it disappears entirely off the screen:

It depends on your menu/window height.
I've removed the min-height so that there is no extra space between the last menu item and the button.

> In collapsed view, the icon is white when I'd expect it to match the color of the other icons:

It seems I'm color blind 😎 . Good catch! Thanks! Fixed !

danielbachhuber reviewed (approved):
Looks great!

One nit: the bottom margin is _slightly_ different for "Add new site" for single user (12px) vs. "Add new site" in Switch Sites (8px):

<img width="583" alt="image" src="https://user-images.githubusercontent.com/36432/189406446-5bf09d5c-78a6-4b2a-9265-74c8633c6581.png">

However, the former is set by `.theme-default .sidebar`, which might be an entire can of worms to change. Given we may be changing the location of "Add new site" in the future, I think we can leave the different in place.

Let's ship this first thing Monday, and then monitor our Tracks event 😊

zaguiini commented on packages/components/src/button/style.scss:
```diff
@@ -287,6 +287,25 @@
 
+.button.is-transparent {
+	background: transparent;
+	border-color: var( --color-sidebar-text );
+	display: flex;
+    justify-content: center;
+
+	&,
+	&:visited {
+		color: var( --color-sidebar-text );
+	}
+
+	&:hover {
+		border-color: var( --color-sidebar-menu-hover-text );
+		color: var( --color-sidebar-menu-hover-text );
```
Let's use different variables here. It's not a good idea to use domain-related variables inside generic components, because these variables might not exist if used outside that specific context/domain.

I suggest using already-known variables, or renaming the variables to something that does not include context:
- `--button-text-color` instead of `--color-sidebar-text`;
- `--button-text-color-hover` instead of `--color-sidebar-menu-hover-text`.

Then, on usage, set the value of such variables in the immediate parent of the button to `var( --color-sidebar-text )` and `var( --color-sidebar-menu-hover-text )`, like:

```css
.immediate-parent-to-transparent-button .transparent-button {
  --button-text-color: var( --color-sidebar-text );
  --button-text-color-hover: var( --color-sidebar-menu-hover-text );
}
```

zaguiini replied:
@sejas, iterating upon my previous comment, here's a diff:

```diff
diff --git a/client/components/site-selector/add-site.tsx b/client/components/site-selector/add-site.tsx
index 478e2fe919..d252c57bcd 100644
--- a/client/components/site-selector/add-site.tsx
+++ b/client/components/site-selector/add-site.tsx
@@ -23,6 +23,7 @@ const SiteSelectorAddSite: FunctionComponent = () => {
 			primary
 			href={ `${ onboardingUrl() }?ref=calypso-selector` }
 			onClick={ recordAddNewSite }
+			className="site-selector__add-new-site"
 		>
 			{ translate( 'Add new site' ) }
 		</Button>
diff --git a/client/components/site-selector/style.scss b/client/components/site-selector/style.scss
index a8df707c13..66e09a19a9 100644
--- a/client/components/site-selector/style.scss
+++ b/client/components/site-selector/style.scss
@@ -229,3 +229,7 @@
 .site-selector__no-results + .site-selector__hidden-sites-message {
 	display: none;
 }
+
+.site-selector__actions :not(.site-selector__add-new-site) {
+	--color-accent: var(--color-sidebar-menu-hover-text);
+}
diff --git a/client/my-sites/sidebar/add-new-site.jsx b/client/my-sites/sidebar/add-new-site.jsx
index 6492ddd5be..fd92e9d052 100644
--- a/client/my-sites/sidebar/add-new-site.jsx
+++ b/client/my-sites/sidebar/add-new-site.jsx
@@ -21,7 +21,12 @@ export const AddNewSite = ( { title } ) => {
 
 	return (
 		<li className="sidebar__actions">
-			<Button transparent href={ `${ onboardingUrl() }?ref=calypso-sidebar` } onClick={ onClick }>
+			<Button
+				className="sidebar__add-new-site-button"
+				transparent
+				href={ `${ onboardingUrl() }?ref=calypso-sidebar` }
+				onClick={ onClick }
+			>
 				<span class="sidebar__action--collapsed dashicons dashicons-plus-alt"></span>
 				<span>{ title }</span>
 			</Button>
diff --git a/client/my-sites/sidebar/style.scss b/client/my-sites/sidebar/style.scss
index c0fb324869..a81b80d32c 100644
--- a/client/my-sites/sidebar/style.scss
+++ b/client/my-sites/sidebar/style.scss
@@ -47,7 +47,11 @@ $font-size: rem(14px);
 	align-items: stretch;
 	justify-content: flex-end;
 	margin-top: auto;
-	padding: 8px 8px 0 8px ;
+	padding: 8px 8px 0 8px;
+}
+
+.sidebar__add-new-site-button {
+	--color-accent: var(--color-sidebar-menu-hover-text);
 }
 
 .sidebar__action--collapsed {
diff --git a/packages/components/src/button/index.stories.jsx b/packages/components/src/button/index.stories.jsx
index 886c4821b8..61de6e1565 100644
--- a/packages/components/src/button/index.stories.jsx
+++ b/packages/components/src/button/index.stories.jsx
@@ -31,7 +31,7 @@ export const Transparent = () => {
 	const sidebarStyle = {
 		...wrapperStyle,
 		gap: '1em',
-		backgroundColor: 'var( --color-sidebar-background )',
+		backgroundColor: 'black',
 	};
 	return (
 		<>
diff --git a/packages/components/src/button/stories.scss b/packages/components/src/button/stories.scss
index 3df43bd5e5..d13167a937 100644
--- a/packages/components/src/button/stories.scss
+++ b/packages/components/src/button/stories.scss
@@ -1,5 +1,3 @@
 .theme-default {
-	--color-sidebar-background: #23282d;
-	--color-sidebar-text: #eee;
-	--color-sidebar-menu-hover-text: #00b9eb;
+	color: #fff;
 }
diff --git a/packages/components/src/button/style.scss b/packages/components/src/button/style.scss
index f0a1d19871..82aa15192b 100644
--- a/packages/components/src/button/style.scss
+++ b/packages/components/src/button/style.scss
@@ -289,18 +289,19 @@
 
 .button.is-transparent {
 	background: transparent;
-	border-color: var( --color-sidebar-text );
+	color: inherit;
+	border-color: currentColor;
 	display: flex;
-    justify-content: center;
+	justify-content: center;
 
 	&,
 	&:visited {
-		color: var( --color-sidebar-text );
+		color: inherit;
 	}
 
 	&:hover {
-		border-color: var( --color-sidebar-menu-hover-text );
-		color: var( --color-sidebar-menu-hover-text );
+		border-color: var(--color-accent);
+		color: var(--color-accent);
 		background: transparent;
 	}
 }
```

Ideally, both the sidebar and the site selector would be redefining `--color-accent` in that context (`.theme-default`), but they're using a new variable called `--color-sidebar-menu-hover-text` for some reason 🤷  This causes us the need to revert to the actual `--color-accent` when needed, just like I did there.

zaguiini replied:
So my suggestion is that we use the current color as the border and text color, and the accent color in the hover state. This feels like the most generic solution possible.

sejas replied:
@zaguiini 
I've added `--button-text-color` and `--button-text-color-hover` CSS variables. I think these names are more descriptive.
But because these variables don't exist for sure, I've added other variables as a fallback. These variables are from the calypso color scheme: https://github.com/Automattic/wp-calypso/tree/trunk/packages/calypso-color-schemes

- Commit: https://github.com/Automattic/wp-calypso/pull/67322/commits/5c75038f17b7afb44073cd638f2f8b6b8cfe50b1#diff-0688ad494526b3b921c4444dc8bc6f63efc4ad74c7c1123b061490cfc57f293a

```CSS
	&,
	&:visited {
		color: var( --button-text-color, currentcolor );
	}
	&:hover {
		border-color: var( --button-text-color-hover, var( --color-primary-40 ) );
		color: var( --button-text-color-hover, var( --color-primary-40 ) );
	}
```

zaguiini replied:
That's fine, too, @sejas! I would go one step further and rename them to `--text-color` and `--text-color-hover`, dropping the `--button` prefix. And then apply those variables to the element itself instead of the parent. That's even better than my diff, honestly xD

zaguiini reviewed (changes requested):
Let's not assume sidebar CSS variables are going to be included wherever the transparent button is used.

sejas committed:
Package Components: refactor transparent button css color variables

sejas committed:
Package Components: refactor transparent button css variables to use var(--color-accent) as a fallback

sejas committed:
Package Components: rename transparent button css variable name
